lib/Dialect/Ibis/IbisOps.cpp

```diff
@@ -141,6 +141,143 @@ LogicalResult ReturnOp::verify() {
+                             << actualType << "'";
+  return success();
+}
+
+// verifies that a local port access (non-nested symbol reference) is valid (the
+// target symbol exists) and that the types of the port and the access match.
+template <typename TSrcOp, typename TTargetOp>
+LogicalResult
+verifyLocalPortSymbolUse(TSrcOp op,
+                         llvm::function_ref<Type(TSrcOp)> getPortType,
+                         SymbolTableCollection &symbolTable) {
+  FlatSymbolRefAttr symName = op.getSymNameAttr();
+  auto getTargetOp = [&](Operation *rootAccessOp) -> FailureOr<TTargetOp> {
+    auto targetOp = dyn_cast_or_null<TTargetOp>(rootAccessOp);
+    if (!targetOp)
+      return op->emitOpError()
+             << "expected '" << symName << "' to refer to a '"
+             << TTargetOp::getOperationName() << "' operation";
+    return {targetOp};
+  };
+
+  return verifyPortSymbolUses<TSrcOp, TTargetOp>(
+      op, symName.getAttr(), getPortType, symbolTable, getTargetOp);
+}
+
+LogicalResult PortReadOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  return verifyLocalPortSymbolUse<PortReadOp, OutputPortOp>(
+      *this, [](PortReadOp op) { return op.getOutput().getType(); },
+      symbolTable);
+}
+
+//===----------------------------------------------------------------------===//
+// PortWriteOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult
+PortWriteOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  return verifyLocalPortSymbolUse<PortWriteOp, InputPortOp>(
+      *this, [](PortWriteOp op) { return op.getInput().getType(); },
+      symbolTable);
+}
+
+//===----------------------------------------------------------------------===//
+// InstanceReadOp
+//===----------------------------------------------------------------------===//
+
+// verifies that an instance port access (nested symbol reference) is valid
+// (the target instance exists, and the target port inside the referenced class
+// exists) and that the types of the port and the access match.
+template <typename TSrcOp, typename TTargetOp>
+LogicalResult
+verifyInstancePortSymbolUses(TSrcOp op,
+                             llvm::function_ref<Type(TSrcOp)> getPortType,
+                             SymbolTableCollection &symbolTable) {
+  auto symName = op.getSymNameAttr();
+  auto getTargetOp = [&](Operation *rootAccessOp) -> FailureOr<TTargetOp> {
+    auto targetInstance = dyn_cast_or_null<InstanceOp>(rootAccessOp);
+    if (!targetInstance)
+      return op->emitOpError()
+             << "expected " << symName.getRootReference() << " to refer to a '"
+             << InstanceOp::getOperationName() << "' operation";
+
+    // Lookup the port in the instance. For now, only allow top level accesses -
+    // can easily extend this to nested instances as well.
+    ClassOp referencedClass = targetInstance.getClass();
+    auto targetOp = symbolTable.lookupSymbolIn<TTargetOp>(
```

Comment:
Here too add a TODO re making this more efficient using innersymtablecollection when that's available to non-firrtl dialects.